tools/codegen/dest/ufunc.py

```diff
@@ -0,0 +1,440 @@
+from dataclasses import dataclass, field
+from typing import Union, Optional, List, Tuple, Dict, Sequence
+from tools.codegen.api.translate import translate
+from tools.codegen.utils import Target
+from tools.codegen.model import NativeFunctionsGroup, ScalarType, UfuncKey, DispatchKey, BaseType, BaseTy, Argument
+import tools.codegen.api.ufunc as ufunc
+from tools.codegen.api.ufunc import UfunctorBindings
+import tools.codegen.api.structured as structured
+from tools.codegen.api.types import StructuredImplSignature, scalar_t, opmath_t, Binding, CType, BaseCType, Expr, NamedCType, ScalarTypeToCppMapping, scalarT, VectorizedCType
+from tools.codegen.context import with_native_function
+from copy import copy
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+#
+#                                  CUDA STUFF
+#
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+
+# NB: not bothering to generate dispatch stub forward declaration in header,
+# we can just paste it whereever necessary
+
+# TODO: use BackendIndex
+# dispatch_key: DispatchKey  # only CPU/CUDA right now
+
+
+# Represents functors for implementing CUDA ufuncs.
+# Functors are templated by scalar_t because when USERS instantiate functors
+# they are templated.  A functor looks something like this:
+#
+#
+@dataclass(frozen=True)
+class UfunctorSignature:
+    g: NativeFunctionsGroup
+    scalar_tensor: Optional[int]
+    name: str
+
+    def arguments(self) -> UfunctorBindings:
+        return ufunc.ufunctor_arguments(self.g, scalar_tensor=self.scalar_tensor, scalar_t=scalar_t)
+
+    def fields(self) -> List[Binding]:
+        # fields are renamed to have a trailing underscore, as is conventional
+        return [b.rename(f"{b.name}_") for b in self.arguments().ctor]
+
+    def returns_type(self) -> CType:
+        # TODO: don't hardcode; return type will be inferred based on tags on
+        # the native function
```

Comment:
when would the return type of the functor's `apply()` function not be a `scalar_t`?

Reply:
(oh, are there some kernels that need the return type to be `opmath_t` to deal with overflow? If it's that, then using tags to figure it out makes sense to me)

Reply:
Unrelated tangent on the subject of tags - I wonder how we want to think about metadata about an op that is:
(1) only useful to codegen
(2) only useful to runtime
(3) useful to both

In the case above, [return type of the functor's apply] sounds useful to codegen, but idk if it's useful to anyone at runtime (maybe we want a way to separate those "types" of tags? Or just not make it a tag in that case, and say that tags are *always* runtime concepts?)